shiny/express/_run.py

```diff
@@ -0,0 +1,243 @@
+            )
+        file = Path(os.getcwd()) / app_file
+
+    # TODO: title and lang
+    app_ui = ui.page_output("__page__")
+
+    def express_server(input: Inputs, output: Outputs, session: Session):
+        try:
+            dyn_ui = run_express(file)
+
+            @render.ui
+            def __page__():
+                return dyn_ui
+
+        except Exception:
+            import traceback
+
+            traceback.print_exception(*sys.exc_info())
+            raise
+
+    app = App(app_ui, express_server)
+
+    return app
+
+
+def is_express_app(app: str, app_dir: str | None) -> bool:
+    """Detect whether an app file is a Shiny express app
+
+    Parameters
+    ----------
+    app
+        App filename, like "app.py". It may be a relative path or absolute path.
+    app_dir
+        Directory containing the app file. If this is `None`, then `app` must be an
+        absolute path.
+
+    Returns
+    -------
+    :
+        `True` if it is a Shiny express app, `False` otherwise.
+    """
+    if not app.lower().endswith(".py"):
+        return False
+
+    if app_dir is not None:
+        app_path = Path(app_dir) / app
+    else:
+        app_path = Path(app)
+
+    if not app_path.exists():
+        return False
+
+    # Read the file, parse it, and look for any imports of shiny.express.
+    with open(app_path) as f:
+        content = f.read()
+    tree = ast.parse(content, app_path)
+    detector = DetectShinyExpressVisitor()
+    detector.visit(tree)
+
+    return detector.found_shiny_express_import
+
+
+def run_express(file: Path) -> Tag | TagList:
+    with open(file) as f:
+        content = f.read()
+
+    tree = ast.parse(content, file)
+    tree = DisplayFuncsTransformer().visit(tree)
+    tree = ast.fix_missing_locations(tree)
+
+    ui_result: Tag | TagList = TagList()
+
+    def set_result(x: object):
+        nonlocal ui_result
+        ui_result = cast(Tag, x)
+
+    sys.displayhook = set_result
+
+    reset_top_level_recall_context_manager()
+    get_top_level_recall_context_manager().__enter__()
+
+    file_path = str(file.resolve())
+
+    var_context: dict[str, object] = {
+        "__file__": file_path,
+        display_decorator_func_name: _display_decorator_function_def,
+    }
+
+    # Execute each top-level node in the AST
+    for node in tree.body:
+        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
+            exec(
+                compile(ast.Module([node], type_ignores=[]), file_path, "exec"),
+                var_context,
+                var_context,
+            )
+        else:
+            exec(
+                compile(ast.Interactive([node], type_ignores=[]), file_path, "single"),
+                var_context,
+                var_context,
+            )
+
+    get_top_level_recall_context_manager().__exit__(None, None, None)
+    return ui_result
+
+
+_top_level_recall_context_manager: RecallContextManager[Tag]
+_top_level_recall_context_manager_has_been_replaced = False
```

Comment:
When you use a `with layout.foo()` function in the wrong setting, you get an unhelpful error message